torch/_dynamo/utils.py

```diff
@@ -1379,3 +1404,37 @@ def tensor_always_has_static_shape(
     if not is_tensor:
         return True, TensorStaticReason.NOT_TENSOR
     return False, None
+
+
+# Note - this could live in shape_env, but then we would need to plumb the
+# config.assume_static_by_default as an input, and that seems a little annoying for little
+# gain.
+def dynamic_dims_from_tensor(
+    e: torch.Tensor, dynamic_ranges: Optional[Dict[int, MinMaxConstraint]]
+) -> List[DIM_DYNAMISM_STATE]:
+    """
+    Given a tensor, returns a list of dimension dynamism states.
+
+    :param e: The input tensor.
+    :type e: torch.Tensor
+    :param dynamic_ranges: A dictionary containing the indices of dynamic dimensions and their corresponding
+    constraints. Defaults to None.
+    :type dynamic_ranges: Optional[Dict[int, MinMaxConstraint]], optional
+    :return: A list of DIM_DYNAMISM_STATE values representing the dynamism state of each dimension in the input tensor.
+    :rtype: List[DIM_DYNAMISM_STATE]
+
+    An invariant is that the length of this list is the number of dimensions of the tensor.
+    If a dimension is marked in dynamic_ranges, it is set as DYNAMIC.
+    Otherwise, it is set to the default dimension state, either DUCK or STATIC depending on the configuration.
+    """
```

Comment:
assert that dynamic_shapes config must be True?

Reply:
Sure thing